src/parser/jobs/sam/modules/Sen.tsx

```diff
@@ -1,98 +1,307 @@
-	static handle = 'sen'
+	static handle = 'Non-Standard Sen Windows'
 
 	@dependency private suggestions!: Suggestions
 	@dependency private kenki!: Kenki
+	@dependency private timeline!: Timeline
 
-	private sen: {[S in SEN]?: boolean} = {}
-	private allowoverwrite: boolean = false
 	private wasted = 0
 
+	private senStateWindows: SenState[] = []
+
+	private get lastSenState(): SenState | undefined {
+		return _.last(this.senStateWindows)
+	}
+
 	protected init() {
-		// Track sen gain
+
 		this.addEventHook(
 			'cast',
-			{by: 'player', abilityId: Object.keys(SEN_ACTIONS).map(Number)},
-			this.onAction,
+			{
+				by: 'player',
+				abilityId: THINGS_WE_WANT_IN_THE_TABLE,
+			},
+			this.checkCastAndPush,
+		)
+
+		this.addEventHook(
+			'combo',
+			{
+				by: 'player',
+				abilityId: SEN_ACTIONS,
+			},
+			this.onSenGen,
+
 		)
 
-		// Death, as well as all Iaijutsu, remove all available sen
-		this.addEventHook('cast', {by: 'player', abilityId: IAIJUTSU}, this.remove)
 		this.addEventHook(
 			'cast',
-			{by: 'player', abilityId: ACTIONS.HAGAKURE.id},
-			this.onHagakure,
+			{
+				by: 'player',
+				abilityId: SEN_REMOVERS,
+			},
+			this.remove,
 		)
-		this.addEventHook('death', {to: 'player'}, this.remove)
+
+		this.addEventHook('death', {to: 'player'}, this.onDeath)
 
 		// Suggestion time~
 		this.addEventHook('complete', this.onComplete)
 	}
 
-	private onAction(event: CastEvent) {
-		const sen = SEN_ACTIONS[event.ability.guid]
+// Handles Sen Gen
+	private onSenGen(event: ComboEvent) {
+		const action = event.ability.guid
+
+		const lastSenState = this.lastSenState
+
+		if (lastSenState != null && lastSenState.end == null) { // The state already exists
+
+			if (event.hasSuccessfulHit === true) {
+
+				switch (action) {
+					case ACTIONS.YUKIKAZE.id:
+					lastSenState.currentSetsu++
+
+					if (lastSenState.currentSetsu > 1) {
+						lastSenState.overwriteSetsus++
+						lastSenState.currentSetsu = 1
+						lastSenState.isOverwrite = true
+					}
+					break
+
+					case ACTIONS.GEKKO.id:
+					case ACTIONS.MANGETSU.id:
+						lastSenState.currentGetsu++
+
+						if (lastSenState.currentGetsu > 1 ) {
+							lastSenState.overwriteGetsus++
+							lastSenState.currentGetsu = 1
+							lastSenState.isOverwrite = true
+						}
+
+						break
+
+					case ACTIONS.KASHA.id:
+					case ACTIONS.OKA.id:
+						lastSenState.currentKa++
+
+						if (lastSenState.currentKa > 1) {
+							lastSenState.overwriteKas++
+							lastSenState.currentKa = 1
+							lastSenState.isOverwrite = true
+						}
+
+						break
+				}
+			}
+		}
+	}
+
+// Function that handles SenState check, if no senState call the maker and then push to the rotation
+	private checkCastAndPush(event: CastEvent) {
+
+		// step 1: set action
+		const action = event.ability.guid
+
+		// step 2: FILTER EVERYTHING
+
+		if (THINGS_WE_WANT_IN_THE_TABLE.includes(action) || SEN_ACTIONS.includes(action) || SEN_REMOVERS.includes(action) ) {
```

Comment:
Redundant check, since we’re applying the first filter in the event hook and the other two filters will never get hit.